.github/workflows/algolia/src/syncAlgoliaIndex.ts

```diff
@@ -0,0 +1,114 @@
+import fs from "fs";
+import path from "path";
+import matter from "gray-matter";
+import algoliasearch from "algoliasearch";
+
+type DocsFrontMatter = {
+  content: string;
+  data: {
+    objectId: string;
+    title?: string;
+    description?: string;
+    slug?: string;
+    tags?: Array<string>;
+  };
+};
+
+const getAllFiles = function (
+  dirPath,
+  arrayOfFiles: Array<{ inputPath: string; outputPath: string; type: string }>,
+  type
+) {
+  const files = fs.readdirSync(dirPath);
+
+  arrayOfFiles = arrayOfFiles || [];
+
+  files.forEach(function (file) {
+    if (fs.statSync(dirPath + "/" + file).isDirectory()) {
+      arrayOfFiles = getAllFiles(dirPath + "/" + file, arrayOfFiles, type);
+    } else {
+      const inputPath = path.join(dirPath, "/", file);
+      const outputPath = `.\\output\\${path.join(
+        "./output/",
+        dirPath,
+        "/",
+        file
+      )}`;
+      arrayOfFiles.push({ inputPath, outputPath, type });
+    }
+  });
+
+  return arrayOfFiles;
+};
+
+const generateAlgoliaJson = () => {
+  const getFormattedData = (matterData: DocsFrontMatter, type) => {
+    const appendData = {
+      ...matterData.data,
+      objectId: undefined,
+      content: matterData.content,
+      objectID: `${type}/${matterData.data.slug}`,
+      type,
+    };
+
+    return appendData;
+  };
+
+  const glossaryFiles = getAllFiles(
+    "../../../resources/glossary",
+    [],
+    "glossary"
+  );
+  const docsFiles = getAllFiles(
+    "../../../resources/docs/simulation",
+    [],
+    "docs"
+  );
+
+  const files = [...glossaryFiles, ...docsFiles];
+
+  const jsonData = files.map((filePath) => {
+    const file = fs.readFileSync(filePath.inputPath, "utf8");
+
+    const grayMatterData = matter(file) as unknown as DocsFrontMatter;
+
+    return getFormattedData(grayMatterData, filePath.type);
+  });
+
+  return jsonData;
+};
+
+export const syncAlgoliaIndex = async () => {
+  const client = algoliasearch(
+    process.env.ALGOLIA_PROJECT,
+    process.env.AGOLIA_WRITE_KEY
+  );
+
+  const index = client.initIndex("hash_learn_testing1");
+
+  let hits: Array<{ objectID: string }> = [];
+
+  await index
+    .browseObjects({
+      query: "", // Empty query will match all records
+      filters: "type:docs OR type:glossary",
+      attributesToRetrieve: ["objectID"],
+      batch: (batch) => {
+        hits = hits.concat(batch);
+      },
+    })
+    .catch(console.error);
+
+  const objectIds = hits.map((hit) => hit.objectID);
+
+  const records = generateAlgoliaJson();
+
+  // delete moved/removed records from index
+  records.forEach((record) => {
+    if (!objectIds.includes(record.objectID)) {
+      index.deleteObject(record.objectID);
+    }
+  });
```

Comment:
Hmm should we check this the other way around? I.e. scan hits search for matching records.

Minor note: the speed of `Array.includes` depends on the length, so it’s slightly better to use a lookup. Even if this does not have direct performance implications today, it might be still worth refactoring. If we do this, folks who see the file for the first time will need to spend less time thinking if they’ve found a performance issue or it’s all fine. Alternatively, we can add a comment saying something about performance (‘we know that `includes` inside `forEach` this generally a bad idea, but we have deal with 1K × 1K so it’s really quick’).

Reply:
By lookup, do you mean that I should convert `objectIds` to an object instead? Any thoughts on what the structure should look like?

Reply:
> Hmm should we check this the other way around? I.e. scan hits search for matching records.

I think it might already be occurring in this way? Have updated variable names to reflect what's happening there: https://github.com/hashintel/hash/pull/49/commits/6d4748bd775937bc3fb5959f499c96985b4371b9#diff-8b8752354bffe5ca09c1f5503d888f007bb97c047e49692b58a80795ec5d189dR107

Reply:
Hmm maybe I'm missing something but here is how I see current logic. Imagine we have objects `A B C` in the server index and `A B D` coming from `generateAlgoliaJson()`. We iterate over `A B D` and notice that `D` is not amongst `A B C`. Then we call `index.deleteObject(D)`, which ends up a noop.

So instead of `generatedRecords.forEach` should it be `hits.forEach`? We can rename `hits` into something like `oldIndexObjects` for more clarity.

As of using a lookup, here is what’s possible to do:

```ts
const indexObjects = generateAlgoliaJson();

const indexObjectLookup: Record<string, YourTypeForIndexObject> = {};
indexObjects.forEach((indexObject) => {
  indexObjectLookup[indexObject.objectID] = indexObject;
});
```

Now we can:

```ts
const objectIDsToDelete: string[] = [];
oldIndexObjects.forEach(({ objectID }) => {
  if (!indexObjectLookup[objectID]) {
    objectIDsToDelete.push(objectID);
  }
});
await index.deleteObjects(objectIDsToDelete);
```

Reply:
Ahh right. Thanks for this, it makes much more sense now.

Reply:
Implemented here: https://github.com/hashintel/hash/pull/49/files#diff-8b8752354bffe5ca09c1f5503d888f007bb97c047e49692b58a80795ec5d189dR100